pkg/apiserver/debugapi/endpoint.go

```diff
@@ -0,0 +1,178 @@
+	Path      string             `json:"path"`
+	Method    string             `json:"method"`
+	Host      EndpointAPIParam   `json:"host"`
+	Segment   []EndpointAPIParam `json:"segment"` // e.g. /stats/dump/{db}/{table} -> db, table
+	Query     []EndpointAPIParam `json:"query"`   // e.g. /debug/pprof?seconds=1 -> seconds
+}
+
+func (e *EndpointAPI) NewRequest(value ValueGetter) (*http.Request, error) {
+	host, err := e.PopulateHost(value)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "populate host")
+	}
+
+	r, err := http.NewRequest(e.Method, fmt.Sprintf("//%s", host), nil)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "new request")
+	}
+
+	// we can put param and value together at first
+	pValues := newParamValues(e.Segment, value)
+	path, err := e.PopulatePath(pValues)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "populate path")
+	}
+	r.URL.Path = path
+
+	qValues := newParamValues(e.Query, value)
+	rawQuery, err := e.EncodeQuery(qValues)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "encode queries")
+	}
+	r.URL.RawQuery = rawQuery
+
+	return r, nil
+}
+
+func (e *EndpointAPI) PopulateHost(value ValueGetter) (string, error) {
+	return e.Host.Transform(value.Get(e.Host.Name))
+}
+
+var ParamRegexp *regexp.Regexp = regexp.MustCompile(`\{(\w+)\}`)
+
+func (e *EndpointAPI) PopulatePath(pvs paramValues) (string, error) {
+	var returnErr error
+	replacedPath := e.Path
+	replacedPath = ParamRegexp.ReplaceAllStringFunc(replacedPath, func(s string) string {
+		if returnErr != nil {
+			return s
+		}
+
+		key := ParamRegexp.ReplaceAllString(s, "${1}")
+		pv, ok := pvs.Find(key)
+		// means the param can be found in the endpoint path, but not in the endpoint param
+		if !ok {
+			returnErr = ErrPathParamNotMatched.New("path: %s, param: %s", e.Path, key)
+			return s
+		}
+
+		val, err := pv.TransformedValue()
+		if err != nil {
+			returnErr = err
+			return s
+		}
+
+		return val
+	})
+	return replacedPath, returnErr
+}
+
+func (e *EndpointAPI) EncodeQuery(pvs paramValues) (string, error) {
+	query := url.Values{}
+	for _, q := range e.Query {
+		// cuz paramValues was generated by e.Query, paramValues can always find param by q.Name
+		pv, _ := pvs.Find(q.Name)
+		val, err := pv.TransformedValue()
+		if err != nil {
+			return "", err
+		}
+		query.Add(q.Name, val)
+	}
+	return query.Encode(), nil
+}
+
+type paramValues map[string]paramValue
+
+func newParamValues(params []EndpointAPIParam, value ValueGetter) paramValues {
+	pvs := paramValues{}
+	for _, p := range params {
+		pvs[p.Name] = paramValue{Param: p, RawValue: value.Get(p.Name)}
+	}
+	return pvs
+}
+
+func (v paramValues) Find(key string) (paramValue, bool) {
+	pv, ok := v[key]
+	return pv, ok
+}
+
+type paramValue struct {
+	Param    EndpointAPIParam
+	RawValue string
+}
+
+func (pv *paramValue) TransformedValue() (string, error) {
+	return pv.Param.Transform(pv.RawValue)
+}
```

Comment:
Is it possible to store the transformed value directly, instead of storing a model and a raw value?